privateSBOMExchange/src/petra/lib/models/__init__.py

```diff
@@ -1,13 +1,304 @@
+    def visit_sbom_node(self, node:SbomNode):
+        """Visit an SbomNode and compute the hash of its data and children.
+
+        The hash is computed using the formula:
+        H( (sbomName) | children)
+
+        Parameters
+        ----------
+        node : SbomNode
+            The SBOM node whose data and children will be hashed.
+
+        Returns
+        -------
+        bytes
+            The computed hash of the SBOM node data and its children.
+        """
+        # Compute hash for the root using its data and the hashes of its children
+
+        children_hashes = b''.join(child.accept(self) for child in node.children)
+        data_to_hash = (node.SBOM_name).encode() + children_hashes
+        node.hash=hashlib.sha256(data_to_hash).digest()
+        return node.hash 
+
+
+class PrintVisitor:
+    """Visitor that prints the data and hash of each node."""
+    def visit_field_node(self, node:FieldNode):
+        print(f"Field:{node.field_name}:{node.field_value}")
+        try:
+            print(f"Hash: {node.hash.hex()}")
+        except AttributeError:
+            print("Hashes have not been calculated, you need to first visit the tree using the Merkle Visitor")
+
+    def visit_complex_node(self, node:ComplexNode):
+        print(f"{node.metadata_type_name}:{node.metadata_type_value}") 
+        try:
+            print(f"Hash: {node.hash.hex()}")
+        except AttributeError:
+            print("Hashes have not been calculated, you need to first visit the tree using the Merkle Visitor")
+
+        for child in node.children:
+            child.accept(self)
+            
+    def visit_sbom_node(self, node:SbomNode):
+        print(f"SBOM: {node.SBOM_name}") 
+        try:
+            print(f"Hash: {node.hash.hex()}")
+        except AttributeError:
+            print("Hashes have not been calculated, you need to first visit the tree using the Merkle Visitor")
+        for child in node.children:
+            child.accept(self)        
+
+
+class EncryptVisitor:
+    """Visitor that encrypts the data in the nodes based on policies."""
+    
+    def __init__(self, policy_file):
+        self.policies = self.load_policies(policy_file)
+
+    def visit_field_node(self, node: FieldNode):
+        """Encrypt the data in a FieldNode."""
+        data_to_encrypt=f"Field{node.field_name}{node.field_value}"
+        #this handle flat fields node that doesnt have complex node as a parent these are document related information
+        if(not node.policy):
+            node.policy = self.get_policy_for_field_node("others",node.field_name)
+        #if parent node(complex node) found policy for this node, encrypt the data using it and erase field node data
+        if node.policy:
+            print(f"policy found for FieldNode '{node.field_name}', {node.policy}.")
+            node.encrypted_data = cpabe_encrypt(pk, node.policy, data_to_encrypt.encode("utf-8"))
+            node.field_name=None
+            node.field_value=None
```

Comment:
If these are supposed to be strings, we should be setting these to the empty string, or some other placeholder string value.